src/dmclone/dm-clone-ioctl.c

```diff
@@ -0,0 +1,183 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <fcntl.h>
+#include <linux/dm-ioctl.h>
+#include <linux/fs.h>
+#include <sys/ioctl.h>
+#include <sys/stat.h>
+
+#include "dm-clone-ioctl.h"
+#include "fd-util.h"
+#include "log.h"
+#include "memory-util.h"         /* memzero() */
+#include "stdio-util.h"          /* xsprintf() */
+#include "string-util.h"
+#include "time-util.h"
+#include "udev-util.h"
+
+static int get_size(const char *dev_path, uint64_t *ret_size) {
+        _cleanup_close_ int fd = -EBADF;
+        uint64_t block_size_bytes;
+
+        assert(dev_path);
+        assert(ret_size);
+
+        fd = open(dev_path, O_RDONLY|O_CLOEXEC);
+        if (fd < 0)
+                return log_error_errno(errno, "Failed to open device '%s': %m", dev_path);
+
+        if (ioctl(fd, BLKGETSIZE64, &block_size_bytes) < 0)
+                return log_error_errno(errno, "Failed to get device size for '%s': %m", dev_path);
+
+        *ret_size = block_size_bytes / 512;
```

Comment:
we have this rule in systemd that we always want to process things with the most basic unit wherever there are multiple possible options. hence, please work in bytes everywhere, and only convert to sectors if the underlying kernel api requires it, right before you call the kernel api. and convert back right after. That simplifies understanding a lot.

in particular as the sector size could be variable here, no?